verifiable-db/src/query/merkle_path.rs

```diff
@@ -291,37 +349,344 @@ where
+        wires
+            .end_node_inputs
+            .left_child_info
+            .set_target(pw, &left_child_info);
+        wires
+            .end_node_inputs
+            .right_child_info
+            .set_target(pw, &right_child_info);
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Enum employed as output type for `build_common` method, which allows to implement the core logic
+/// of `MerklePathGadget` and `MerklePathGadgetWithNeighbors` in a single method, hence avoiding
+/// code duplication
+enum MerklePathWires<const MAX_DEPTH: usize>
+where
+    [(); MAX_DEPTH - 1]:,
+{
+    MerklePath(MerklePathTarget<MAX_DEPTH>),
+    MerklePathWithNeighbors(MerklePathWithNeighborsTarget<MAX_DEPTH>),
+}
+
+/// Build `MerklePathWires`, building `MerklePathWithNeighborsTarget` if the input flag `with_neighbors`
+/// is true, building `MerkelPathTarget` otherwise. The required inputs are:
+/// - `end_node`: The hash of the first node in the path
+/// - `index_id`: Integer identifier of the index column to be placed in the hash
+///     of the nodes of the path
+/// - `end_node_info` : Data about `end_node`; must be provided only if `with_neighbors == true`
+fn build_common<const MAX_DEPTH: usize>(
+    b: &mut CircuitBuilder<F, D>,
+    end_node: HashOutTarget,
+    index_id: Target,
+    with_neighbors: bool,
+    end_node_info: Option<EndNodeInputs>,
+) -> MerklePathWires<MAX_DEPTH>
+where
+    [(); MAX_DEPTH - 1]:,
+{
+    let is_left_child = array::from_fn(|_| b.add_virtual_bool_target_unsafe());
+    let [sibling_hash, embedded_tree_hash] =
+        [0, 1].map(|_| array::from_fn(|_| b.add_virtual_hash()));
+    let [node_min, node_max, node_value] = [0, 1, 2].map(
+        |_| b.add_virtual_u256_arr_unsafe(), // unsafe should be ok since we just need to hash them
+    );
+    let is_real_node = array::from_fn(|_| b.add_virtual_bool_target_safe());
+
+    let (mut predecessor_info, mut successor_info) = if with_neighbors {
+        // we need to initialize predecessor and successor data
+        let end_node_info = end_node_info.as_ref().unwrap();
+        // the predecessor of end_node is an ancestor of end_node iff end_node has no left child
+        let is_predecessor_in_path = b.not(end_node_info.left_child_exists);
+        let zero_u256 = b.zero_u256();
+        // Initialize value of predecessor node of end_node to a dummy value if the predecessor node
+        // will be found in the path; otherwise, the predecessor_value is the maximum value in
+        // the subtree rooted in the left child of end_node
+        let predecessor_value = b.select_u256(
+            is_predecessor_in_path,
+            &zero_u256,
+            &end_node_info.left_child_info.max,
+        );
+        // the predecessor value is already found if end_node has a left child
+        let predecessor_found = end_node_info.left_child_exists;
+        // Initialize predecessor node hash to a dummy value
```

Comment:
can you tell why ? Especially when the predecessor is within the left subtree, when this would be set to the right hash ?